contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,471 @@
+
+The scheduler is still be in charge of filtering the nodes which cannot
+satisfy the resource requests.
+He might in the future be in charge of selecting the device.
+
+## Device Plugin
+
+### Introduction
+The device plugin is structured in 5 parts:
+1. Registration: The device plugin advertises it's presence to Kubelet
+2. Discovery: Kubelet calls the device plugin to list it's devices
+3. Allocate / Deallocate: When creating/deleting containers requesting the
+   devices advertised by the device plugin, Kubelet calls the device plugin's
+   `Allocate` and `Deallocate` functions.
+4. Cleanup: Kubelet terminates the communication through a "Stop"
+4. Heartbeat: The device plugin polls Kubelet to know if it's still alive
+   and if it has to re-issue a Register request
+
+### Registration
+
+When starting the device plugin is expected to make a (client) gRPC call
+to the `Register` function that Kubelet exposes.
+
+The communication between Kubelet is expected to happen only through Unix
+sockets and follow this simple pattern:
+1. The device plugins starts it's gRPC server
+2. The device plugins sends a `RegisterRequest` to Kubelet (through a
+   gRPC request)
+4. Kubelet starts it's Discovery phase and calls `Discover` and `Monitor`
+5. Kubelet answers to the `RegisterRequest` with a `RegisterResponse`
+   containing any error Kubelet might have encountered
+
+### Unix Socket
+
+Device Plugins are expected to communicate with Kubelet through gRPC
+on an Unix socket.
+When starting the gRPC server, they are expected to create a unix socket
+at the following host path: `/var/run/Kubernetes`.
+
+For non bare metal device plugin this means they will have to mount the folder
+as a volume in their pod spec ([see Installation](##installation)).
+
+Device plugins can expect to find the socket to register themselves on
+the host at the following path:
+`/var/run/Kubernetes/kubelet.sock`.
+
+### Protocol Overview
+
+When first registering themselves against Kubelet, the device plugin
+will send:
+  * The name of their unix socket
+  * [The API version against which they were built](#versioning).
+  * Their `Vendor` ID or name of the device plugin
+
+Kubelet answers with the minimum version it supports and whether or
+not there was an error. The errors may include (but not limited to):
+  * API version not supported
+  * A device plugin was already registered for this vendor
+  * A device plugin already registered this device
+  * Vendor is not consistent across discovered devices
+
+Kubelet will then interact with the plugin through the following functions:
+  * `Discover`: List Devices
+  * `Monitor`: Returns a stream that is written to when a
+     Device becomes unhealty
+  * `Allocate`: Called when creating a container with a list of devices
+     can request changes to the Container config
+  * `Deallocate`: Called when deleting a container can be used for cleanup
+
+The device plugin is also expected to periodically call the `Heartbeat` function
```

Comment:
The heartbeat is more of a mechanism for the Device Plugins to make sure that Kubelet is alive rather than a mechanism for Kubelet to make sure that the Device Plugins are alive